app/javascript/articles/components/SaveButton.jsx

```diff
@@ -33,7 +34,14 @@ export class SaveButton extends Component {
       });
     };
 
-    if (article.class_name === 'Article') {
+    const articleOwnedByCurrentUser = function (article) {
+      return article.user_id === userData().id;
```

Comment:
I'm using the `id` value from `userData()` to check if the post belongs to the current user.

However, this approach is giving me issue in the test (see below). From multiple test suites (that probably use the <Article /> component), I keep getting the error:
```javascript
TypeError: Cannot read properties of null (reading 'id')

      36 |
      37 |     const articleOwnedByCurrentUser = function (article) {
    > 38 |     return article.user_id === userData().id;
         |                                  ^
      39 |     };
      40 |
      41 |     if (
```

I have tried to mock the `userData()` in the test, but I do not believe I am doing it right.

I'd appreciate help 🙏🏾  @aitchiss

Reply:
The problem we're having here is that `userData` is only initialized when the overall page loads, whereas component level tests test the component in isolation (so none of the code initializing userData is running).

In terms of a solution, I don't think we want to access `userData` inside the `SaveButton` component - it means we are going in and grabbing this data for every article on the feed, when we only really need to access it once. Instead, I would access it in `homePageFeed.jsx` and pass a prop through to the article e.g. something like

```jsx
const { id: userId } = userData();

<Article
  {...commonProps}
  key={story.id}
  article={story}
  feedStyle={feedStyle}
  isBookmarked={bookmarkedFeedItems.has(story.id)}
  saveable={story.user_id !== userId}
/>
```

and then the `Article` can just completely skip rendering the `SaveButton` if the `saveable` prop is `false`.

The testing then becomes more isolated and straightforward, because you can test the Article component doesn't have a save button when the prop is `false`, rather than having to mock userData etc. A snapshot test would be simplest for this, e.g.

```jsx
  it("should render an article that isn't saveable", () => {
    const { container } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={article}
        currentTag="javascript"
        saveable={false}
      />,
    );

    expect(container).toMatchSnapshot();
  });
```

Reply:
Thank you so so so much for this! 🙏🏾